# ViralWasm-Epi 

A client-side WebAssembly pipeline for multi-sequence alignment and molecular clustering.

Visit: https://niema-lab.github.io/ViralWasm-Epi/ to run the pipeline.

## ViralWasm-Epi Offline
**Prerequisites**: Python (Can be downloaded here: https://www.python.org/downloads/)

To run the pipeline locally without internet, download the zip file containing the offline mode of ViralWasm-Epi: https://github.com/Niema-Lab/ViralWasm-Epi/raw/master/viralwasm-epi-offline.zip. After, extract the folder, and click on the `run_website.py` file and select "Open with Python" for Windows or "Open With -> Python Launcher" for Mac. 


Note: When starting the website for the first time offline, you may be prompted with a firewall warning. Click "Allow Access" to run the pipeline.

## Citing ViralWasm-Epi

If you use ViralWasm-Epi in your work, please cite:

> Ji D, Aboukhalil R, Moshiri N (2023). "ViralWasm: a client-side user-friendly web application suite for viral genomics." *Bioinformatics*. btae018. [doi:10.1093/bioinformatics/btae018](https://doi.org/10.1093/bioinformatics/btae018)

Please also cite the following tools used in ViralWasm-Epi:

### Biowasm

See https://biowasm.com/.

### Pyodide

See https://github.com/pyodide/pyodide.

### ViralMSA

> Moshiri N (2021). "ViralMSA: massively scalable reference-guided multiple sequence alignment of viral genomes." *Bioinformatics*. 37(5):714–716. [doi:10.1093/bioinformatics/btaa743](https://doi.org/10.1093/bioinformatics/btaa743)

### Minimap2

> Li H (2018). "Minimap2: pairwise alignment for nucleotide sequences." *Bioinformatics*. 34(18):3094–3100. [doi:10.1093/bioinformatics/bty191](https://doi.org/10.1093/bioinformatics/bty191)

### TN93 

See https://github.com/veg/tn93. 

> Tamura K, Nei M (1993). "Estimation of the number of nucleotide substitutions in the control region of mitochondrial DNA in humans and chimpanzees." *Molecular Biology and Evolution*. 10(3):512–526. [doi:10.1093/oxfordjournals.molbev.a040023](https://doi.org/10.1093/oxfordjournals.molbev.a040023)

### FastTree

> Price MN, Dehal PS, Arkin AP (2010). "FastTree 2 – Approximately Maximum-Likelihood Trees for Large Alignments." *PLOS ONE*. 5(3):e9490. [doi:10.1371/journal.pone.0009490](https://doi.org/10.1371/journal.pone.0009490)

### Seqtk

See https://github.com/lh3/seqtk.

### LSD2

> To TH, Jung M, Lycett S, Gascuel O (2016). "Fast dating using least-squares criteria and algorithms." *Systematic Biology*. 65(1):82–97. [doi:10.1093/sysbio/syv068](https://doi.org/10.1093/sysbio/syv068)
